File: Ruby/lib/items/normal_item.rb
# frozen_string_literal: true

require 'items/item'

module Items
  # Represents items whose quality gradually decreases to 0 over time
  class NormalItem < Item
    def tick
      @gilded_rose.days_remaining -= 1

      return if @gilded_rose.quality <= 0

      @gilded_rose.quality -= 1
      @gilded_rose.quality -= 1 if @gilded_rose.days_remaining <= 0
      ensure_bounded_quality
    end
  end
end

File: JavaScript/src/items/NormalItem.js
import Item from './Item';

// Represents items whose quality gradually decreases to 0 over time
export default class NormalItem extends Item {
  tick() {
    this.gildedRose.daysRemaining -= 1;

    if (this.gildedRose.quality <= 0) {
      return;
    }

    this.gildedRose.quality -= 1;
    if (this.gildedRose.daysRemaining <= 0) {
      this.gildedRose.quality -= 1;
    }

    this.ensureBoundedQuality();
  }
}
